src/eko/evolution_operator/c_quad_ker.cpp

```diff
@@ -0,0 +1,149 @@
+    ekorepp::cmplx n() const {
+        const double theta = this->theta();
+        // treat singular point separately
+        const double re = 0.5 == t ? 1.0 : theta / tan(theta);
+        const double im = theta;
+        return o + r * ekorepp::cmplx(re, im);
+    }
+
+    /** @brief transformation jacobian */
+    ekorepp::cmplx jac() const {
+        const double theta = this->theta();
+        // treat singular point separately
+        const double re = 0.5 == t ? 0.0 : 1.0 / tan(theta) - theta / pow(sin(theta), 2);
+        const double im = 1.0;
+        return r * M_PI * 2.0 * ekorepp::cmplx(re, im);
+    }
+
+    /** @brief Mellin inversion prefactor */
+    ekorepp::cmplx prefactor() const {
+        return ekorepp::cmplx(0,-1./M_PI);
+    }
+};
+
+/** @brief Python callback signature  */
+typedef double (*py_quad_ker_qcd)(
+                        double* re_gamma, double* im_gamma,
+                        double re_n, double im_n,
+                        double re_jac, double im_jac,
+                        unsigned int order_qcd,
+                        bool is_singlet,
+                        unsigned int mode0,
+                        unsigned int mode1,
+                        unsigned int nf,
+                        bool is_log,
+                        double logx,
+                        double* areas,
+                        unsigned int areas_x,
+                        unsigned int areas_y,
+                        double L,
+                        unsigned int method_num,
+                        double as1, double as0,
+                        unsigned int ev_op_iterations, unsigned int ev_op_max_order_qcd,
+                        unsigned int sv_mode_num, bool is_threshold);
+
+/** @brief additional integration parameters */
+struct QuadQCDargs {
+    unsigned int order_qcd;
+    unsigned int mode0;
+    unsigned int mode1;
+    bool is_polarized;
+    bool is_time_like;
+    unsigned int nf;
+    py_quad_ker_qcd py;
+    bool is_log;
+    double logx;
+    double* areas;
+    unsigned int areas_x;
+    unsigned int areas_y;
+    double L;
+    unsigned int method_num;
+    double as1;
+    double as0;
+    unsigned int ev_op_iterations;
+    unsigned int ev_op_max_order_qcd;
+    unsigned int sv_mode_num;
+    bool is_threshold;
+};
+
+/**
+ * @brief intergration kernel inside quad
+ * @param u interation variable
+ * @param rargs raw args, to be casted to QuadQCDargs
+ * @return intergration value
+ */
+extern "C" double c_quad_ker_qcd(const double u, void* rargs) {
+    QuadQCDargs args = *(QuadQCDargs* )rargs;
+    const bool is_singlet = (100 == args.mode0) || (21 == args.mode0) || (90 == args.mode0);
+    // prepare gamma
+    const TalbotPath path(u, args.logx, is_singlet);
+    const ekorepp::cmplx jac = path.jac() * path.prefactor();
+    ekorepp::harmonics::Cache c(path.n());
+    const size_t size = args.order_qcd * (is_singlet ? 4 : 1);
+    std::vector<double> re(size);
```

Comment:
why do you need to split real and imaginary part, and then recombine ?

Reply:
I believe you can not bind complex between C++ and Python (at least not with `ctypes` https://docs.python.org/3/library/ctypes.html#ctypes.c_byte).

For many languages, complex numbers are not basic types. Even in C++, despite being in standard library, they are not as primitive as floats and doubles (I believe the main reason is that you do not have native instructions to work with complex in most common architectures).

Reply:
indeed